driver-async/src/main/com/mongodb/async/client/gridfs/GridFS.java

```diff
@@ -0,0 +1,205 @@
+                    chunkCollection.createIndex(new Document("files_id", 1).append("n", 1),
+                            new CreateIndexOptions().unique(true), new SingleResultCallback<Void>() {
+
+                        @Override
+                        public void onResult(final Void result, final Throwable t) {
+                            //TODO log
+                            if (t != null) {
+                                //failed to create index
+                            } else {
+                                // index ok
+                            }
+                        }
+                    });
+                }
+            }
+        });
+    }
+
+    /**
+     * Instantiate a new GridFSFile.
+     *
+     * @return GridFSFile
+     */
+    public GridFSFile create() {
+        GridFSFile gridFSFile = new GridFSFile(chunkSize);
+        gridFSFile.setId(new ObjectId());
+        gridFSFile.setUploadDate(new Date());
+        return injectContext(gridFSFile);
+    }
+
+    /**
+     * Find a GridFSFile by its id.
+     *
+     * @param id the id as ObjectId
+     * @param callback callback that is completed when the file is available
+     */
+    public void find(final ObjectId id, final SingleResultCallback<GridFSFile> callback) {
+        fileCollection.find(new Document("_id", id)).first(new SingleResultCallback<GridFSFile>() {
+            @Override
+            public void onResult(final GridFSFile result, final Throwable t) {
+                if (t != null) {
+                    callback.onResult(null, t);
+                } else {
+                    callback.onResult(injectContext(result), null);
+                }
+            }
+        });
+    }
+
+    /**
+     * List all GridFSFiles, restricted by the provided query.
+     * @param query a document to limit the results that are returned
+     * @param callback callback that will be completed when the result is available
+     */
+    public void listFiles(final Document query, final SingleResultCallback<List<GridFSFile>> callback) {
+        fileCollection.find(query)
+                .sort(new Document("filename", 1))
+                .into(new ArrayList<GridFSFile>(), new SingleResultCallback<ArrayList<GridFSFile>>() {
+
+            @Override
+            public void onResult(final ArrayList<GridFSFile> result, final Throwable t) {
+                if (t != null) {
+                    callback.onResult(null, t);
+                } else {
+                    for (GridFSFile file : result) {
+                        injectContext(file);
+                    }
+                    callback.onResult(result, null);
+                }
+            }
+        });
+    }
+
+    /**
+     * Delete a GridFSFile by its id.
+     *
+     * @param id the id as ObjectId
+     * @param callback callback that is completed when the file is deleted
+     */
+    public void delete(final ObjectId id, final SingleResultCallback<DeleteResult> callback) {
```

Comment:
Hm, that currently appears to be consistent with the MongoCollection interface:

````java
void deleteOne(Object filter, SingleResultCallback<DeleteResult> callback);
````

Providing the DeleteResult allows to answer the question "Did I actually delete a file?", which might be useful in some cases. Returning Void could mean to things: 
- one file has been deleted
- no file has been deleted
 
I favour keeping the DeleteResult in this case.

For acknowledged writes we could check if the correct number of chunks has been deleted as well and return an error if not? However if such a operation fails, this usually leaves the database in an inconsistent state. I was wondering if we should provide an option to mark the files document as being deleted before issuing any actual delete operation, delete the chunks first and then finally delete the files document. That at least would make it easy to find any GridFS files where anything went wrong.